cpp/include/cudf/types.hpp

```diff
@@ -136,6 +136,15 @@ enum class nan_policy : bool {
   NAN_IS_VALID  ///< treat nans as valid elements (non-null)
 };
 
+/**
+ * @brief Enum to consider different elements (of floating point types) holding NaN value as equal
+ * or unequal
+ */
+enum class nan_equality /*unspecified*/ {
+  ALL_EQUAL,  ///< All NaNs compare equal, regardless of sign
+  UNEQUAL     ///< All NaNs compare unequal (IEE754 behavior)
```

Comment:
FWIW, I like the naming of `ALL_EQUAL`. And maybe the naming of `UNEQUAL` should be `ALL_UNEQUAL`?
And the docstrings should be more explicit:
```
ALL_EQUAL --> +NaN == +NaN && +NaN == -NaN && -NaN == -NaN,
ALL_UNEQUAL --> +NaN != +NaN && +NaN != -NaN && -NaN != -NaN
```

Side Q: Doesn't IEEE754 call the two nans `sNaN` and `qNaN`?

Reply:
All `NaNs` compare unequall---this should be enough to cover every case: every `NaN` is unique.

`sNaN` and `qNaN` are also different. `sNaN` is never the result of the computation, it only exists if we explicitly add it to the data. I'll add it to the tests too.

Reply:
I concede - the docstring is sufficient and gives a clear stance on how libcudf would like to view NaNs (self-consistently).

Reply:
Besides typo `IEE754` --> `IEEE754` ;)